src/compiler/crystal/repl/compiler.cr

```diff
@@ -0,0 +1,3221 @@
+        reference_is_a(type_id(filtered_type), node: node)
+      end
+    when NilableProcType
+      # Remove the closure data
+      pop sizeof(Void*), node: nil
+
+      if filtered_type.nil_type?
+        pointer_is_null(node: node)
+      else
+        pointer_is_not_null(node: node)
+      end
+    when ReferenceUnionType
+      case filtered_type
+      when NonGenericClassType
+        reference_is_a(type_id(filtered_type), node: node)
+      when GenericClassInstanceType
+        # TODO: not tested
+        reference_is_a(type_id(filtered_type), node: node)
+      when VirtualType
+        # TODO: not tested
+        reference_is_a(type_id(filtered_type), node: node)
+      when ReferenceUnionType
+        # TODO: not tested
+        reference_is_a(type_id(filtered_type), node: node)
+      else
+        node.raise "BUG: missing filter type from #{type} to #{filtered_type} (#{type.class} to #{filtered_type.class})"
+      end
+    when VirtualMetaclassType
+      case filtered_type
+      when MetaclassType
+        metaclass_is_a(type_id(filtered_type), node: node)
+      when VirtualMetaclassType
+        metaclass_is_a(type_id(filtered_type), node: node)
+      else
+        node.raise "BUG: missing filter type from #{type} to #{filtered_type} (#{type.class} to #{filtered_type.class})"
+      end
+    else
+      node.raise "BUG: missing filter type from #{type} to #{filtered_type} (#{type.class} to #{filtered_type.class})"
+    end
+  end
+
+  def visit(node : Call)
+    obj = node.obj
+    with_scope = node.with_scope
+
+    if !obj && with_scope && node.uses_with_scope?
+      obj = Var.new(WITH_SCOPE, with_scope)
+    end
+
+    target_defs = node.target_defs
+    unless target_defs
+      node.raise "BUG: no target defs"
+    end
+
+    if target_defs.size == 1
+      target_def = target_defs.first
+    else
+      target_def = Multidispatch.create_def(@context, node, target_defs)
+    end
+
+    body = target_def.body
+    if body.is_a?(Primitive)
+      visit_primitive(node, body)
+      return false
+    end
+
+    if body.is_a?(InstanceVar)
+      # Inline the call, so that it also works fine when wanting to take a pointer through things
+      # (this is how compiled Crystal works too
+      if obj
+        compile_read_instance_var(node, obj, body.name, owner: target_def.owner)
+      else
+        compile_instance_var(body)
+      end
+
+      # We still have to accept the call arguments, but discard their values
+      node.args.each { |arg| discard_value(arg) }
+
+      return false
+    end
+
+    if body.is_a?(Var) && body.name == "self"
+      # We also inline calls that simply return "self"
+
+      if @wants_value
+        if obj
+          request_value(obj)
+        else
+          put_self(node: node)
+        end
+      end
+
+      # We still have to accept the call arguments, but discard their values
+      node.args.each { |arg| discard_value(arg) }
+
+      return false
+    end
```

Comment:
I suppose more calls should be inlined. For example methods annotaed by `@[AlwaysInline]`.
Another question is whether to inline more agressively as LLVM does to keep closer to compiled behaviour (for example with private methods). And of course, determine how much agressively.
This is not necessary immediately, but maybe we could at least consider leaving some comments here. Maybe we could refactor this code to be more generic, which makes adding more inlining cases easy.

Reply:
I don't think so. The reason why there's some inlining here is because if we don't do that, the interpreter wouldn't behave exactly like compiled crystal. I'll show some cases later on. But this was one of the trickiest things when implementing the interpreter, in order to match the compiler's behavior. Unfortunately when I decided to inline these things in compiled Crystal I defined some semantics that I didn't expect.

Inlining other things in the interpreter is pretty hard! I have no idea how to do that. Also, `@[AlwaysInline]` is just a suggestion for LLVM, LLVM can later do whatever it wants, even completely ignore that annotation. And the interpreter will do that: completely ignore it, always.